File: packages/frontend/src/services/facilityService.ts
import type {
  Facility,
  CreateFacilityData,
  UpdateFacilityData,
  TimeBlock,
  ScheduleException,
} from "@/types/facility";
import { apiClient } from "@/lib/apiClient";

/**
 * Create a new facility
 */
export async function createFacility(
  stableId: string,
  facilityData: CreateFacilityData,
  userId: string,
): Promise<string> {
  const response = await apiClient.post<{ id: string }>("/facilities", {
    stableId,
    ...facilityData,
  });

  return response.id;
}

/**
 * Get facility by ID
 */
export async function getFacility(
  facilityId: string,
): Promise<Facility | null> {
  try {
    return await apiClient.get<Facility>(`/facilities/${facilityId}`);
  } catch (error: any) {
    if (error.status === 404) {
      return null;
    }
    throw error;
  }
}

/**
 * Get all facilities for a stable
 */
export async function getFacilitiesByStable(
  stableId: string,
  options?: { reservableOnly?: boolean },
): Promise<Facility[]> {
  const params: Record<string, string> = { stableId };

  if (options?.reservableOnly) {
    params.reservableOnly = "true";
  }

  const response = await apiClient.get<{ facilities: Facility[] }>(
    "/facilities",
    params,
  );

  return response.facilities;
}

/**
 * Get active facilities for a stable
 */
export async function getActiveFacilities(
  stableId: string,
): Promise<Facility[]> {
  const response = await apiClient.get<{ facilities: Facility[] }>(
    "/facilities",
    { stableId, status: "active" },
  );

  return response.facilities;
}

/**
 * Update facility
 */
export async function updateFacility(
  facilityId: string,
  updates: UpdateFacilityData,
  userId: string,
): Promise<void> {
  await apiClient.patch(`/facilities/${facilityId}`, updates);
}

/**
 * Delete facility
 */
export async function deleteFacility(
  facilityId: string,
  _userId?: string,
): Promise<void> {
  await apiClient.delete(`/facilities/${facilityId}`);
}

/**
 * Get available time slots for a facility on a specific date
 */
export async function getAvailableSlots(
  facilityId: string,
  date: string,
): Promise<{ date: string; timeBlocks: TimeBlock[] }> {
  return apiClient.get(`/facilities/${facilityId}/available-slots`, { date });
}

/**
 * Add a schedule exception to a facility
 */
export async function addScheduleException(
  facilityId: string,
  exception: {
    date: string;
    type: "closed" | "modified";
    timeBlocks?: TimeBlock[];
    reason?: string;
  },
): Promise<{ success: boolean; exception: ScheduleException }> {
  return apiClient.post(`/facilities/${facilityId}/exceptions`, exception);
}

/**
 * Remove a schedule exception
 */
export async function removeScheduleException(
  facilityId: string,
  date: string,
): Promise<void> {
  await apiClient.delete(`/facilities/${facilityId}/exceptions/${date}`);
}

File: EquiDuty/EquiDuty/Features/Facilities/Services/FacilityService.swift
//
//  FacilityService.swift
//  EquiDuty
//
//  Implementation of facility operations using API client
//

import Foundation

/// Facility service implementation
@MainActor
final class FacilityService: FacilityServiceProtocol {
    static let shared = FacilityService()

    private let apiClient = APIClient.shared
    private let dateFormatter: DateFormatter

    private init() {
        dateFormatter = DateFormatter()
        dateFormatter.dateFormat = "yyyy-MM-dd"
        dateFormatter.locale = Locale(identifier: "en_US_POSIX")
    }

    func getFacilities(stableId: String, reservableOnly: Bool = false) async throws -> [Facility] {
        var params = ["stableId": stableId]

        if reservableOnly {
            params["reservableOnly"] = "true"
        }

        let response: FacilitiesResponse = try await apiClient.get(
            APIEndpoints.facilities,
            params: params
        )
        return response.facilities
    }

    func getFacility(id: String) async throws -> Facility {
        let response: Facility = try await apiClient.get(
            APIEndpoints.facility(id)
        )
        return response
    }

    func getAvailableSlots(facilityId: String, date: Date) async throws -> AvailableSlotsResponse {
        let dateString = dateFormatter.string(from: date)
        let params = ["date": dateString]
        let response: AvailableSlotsResponse = try await apiClient.get(
            APIEndpoints.facilityAvailableSlots(facilityId),
            params: params
        )
        return response
    }

    func createFacility(_ request: CreateFacilityRequest) async throws -> String {
        let response: CreateFacilityResponse = try await apiClient.post(
            APIEndpoints.facilities,
            body: request
        )
        return response.id
    }

    func updateFacility(id: String, updates: UpdateFacilityRequest) async throws -> Facility {
        let response: Facility = try await apiClient.patch(
            APIEndpoints.facility(id),
            body: updates
        )
        return response
    }

    func deleteFacility(id: String) async throws {
        try await apiClient.delete(APIEndpoints.facility(id))
    }

    func addScheduleException(facilityId: String, exception: CreateScheduleExceptionRequest) async throws -> ScheduleException {
        let response: ScheduleExceptionResponse = try await apiClient.post(
            APIEndpoints.facilityExceptions(facilityId),
            body: exception
        )
        return response.exception
    }

    func removeScheduleException(facilityId: String, date: String) async throws {
        try await apiClient.delete(APIEndpoints.facilityException(facilityId, date: date))
    }
}
